File: C_Programming/Practise_Programs/PP_013.c
//////////////////////////////////////////////////////////////////
//
//  File Name    : Check_Even_Odd
//
//  Description  : This program accepts a number
//                 from the user and checks whether
//                 it is even or odd.
//
//  Example:
//      Input  : 20
//      Output : 20 Number is Even
//
//////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////
//
//  Enhanced Version: Added boolean return type function and
//                    displays both input number and Even/Odd result.
//
//////////////////////////////////////////////////////////////////

/*
    Algorithm

    START

        Accept number as No

        Call CheckEvenOdd() by passing No

        Receive the result

        If the result is true
            Display that No is Even

        Otherwise
            Display that No is Odd

    STOP
*/

#include<stdio.h>
#include<stdbool.h>

bool CheckEvenOdd(int iNo)
{
    int iRemainder = 0;

    iRemainder = iNo % 2;

    if(iRemainder == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

int main()
{
    int iValue = 0;
    bool bRet = false;
    
    printf("Enter number to checked whether it is Even or Odd : ");
    scanf("%d", &iValue);
   
    bRet = CheckEvenOdd(iValue);

    if(bRet == true)
    {
        printf("%d Number is Even\n", iValue);
    }
    else
    {
        printf("%d Number is Odd\n", iValue);
    }
    return 0;
}

File: Java_Programming/Practise_Programs/PP_013.java
//////////////////////////////////////////////////////////////////
//
//  File Name    : PP_013
//
//  Description  : This program accepts a number from the user 
//                 and displays all of its factors except the number
//                 itself using an optimized approach that 
//                 iterates up to half of the number.
//
//  Example:
//      Input  : 12
//      Output : 1
//               2
//               3
//               4
//               6
//
//////////////////////////////////////////////////////////////////

import java.util.Scanner;

class NumberX
{
    public void DisplayFactors(int iNo)
    {
        int iCnt = 0;

        for(iCnt = 1; iCnt <= (iNo / 2); iCnt++)
        {
            if((iNo % iCnt) == 0)
            {
                System.out.println(iCnt);
            }
        }
    }    
}

class PP_013
{
    public static void main(String A[])
    {
        Scanner sobj = new Scanner(System.in);

        int iValue = 0;
        
        System.out.println("Enter number : ");
        iValue = sobj.nextInt(); 
        
        NumberX nobj = new NumberX();

        nobj.DisplayFactors(iValue);
        
    }
}

// Time Complexity : O(N/2)
// Where N >= 0